pkg/maputil/maputil.go

```diff
@@ -60,32 +63,133 @@ func recursivelyStringifyMapKey(v interface{}) (interface{}, error) {
 	return casted_v, nil
 }
 
-func Set(m map[string]interface{}, key []string, value string) map[string]interface{} {
-	if len(key) == 0 {
-		panic(fmt.Errorf("bug: unexpected length of key: %d", len(key)))
-	}
+type arg interface {
+	getMap(map[string]interface{}) map[string]interface{}
+	set(map[string]interface{}, string)
+}
 
-	k := key[0]
+type keyArg struct {
+	key string
+}
 
-	if len(key) == 1 {
-		m[k] = value
-		return m
+func (a keyArg) getMap(m map[string]interface{}) map[string]interface{} {
+	_, ok := m[a.key]
+	if !ok {
+		m[a.key] = map[string]interface{}{}
+	}
+	switch t := m[a.key].(type) {
+	case map[string]interface{}:
+		return t
+	default:
+		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
 	}
+}
 
-	remain := key[1:]
+func (a keyArg) set(m map[string]interface{}, value string) {
+	m[a.key] = value
+}
+
+type indexedKeyArg struct {
+	key   string
+	index int
+}
 
-	nested, ok := m[k]
+func (a indexedKeyArg) getArray(m map[string]interface{}) []interface{} {
+	_, ok := m[a.key]
 	if !ok {
-		nested = map[string]interface{}{}
+		m[a.key] = make([]interface{}, a.index+1)
+	}
+	switch t := m[a.key].(type) {
+	case []interface{}:
+		if len(t) <= a.index {
+			t2 := make([]interface{}, a.index+1)
+			copy(t, t2)
+			t = t2
+		}
+		return t
+	default:
+		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
+	}
+}
+
+func (a indexedKeyArg) getMap(m map[string]interface{}) map[string]interface{} {
+	t := a.getArray(m)
+	if t[a.index] == nil {
+		t[a.index] = map[string]interface{}{}
 	}
-	switch t := nested.(type) {
+	switch t := t[a.index].(type) {
 	case map[string]interface{}:
-		nested = Set(t, remain, value)
+		return t
 	default:
 		panic(fmt.Errorf("unexpected type: %v(%T)", t, t))
 	}
+}
+
+func (a indexedKeyArg) set(m map[string]interface{}, value string) {
+	t := a.getArray(m)
+	t[a.index] = value
+	m[a.key] = t
+}
+
+func getCursor(key string) arg {
```

Comment:
Your usage of the strategy pattern here was eye-opening to me! If it was me I would have written this in switch-case nested in a for-loop, but yours is MUCH more readable. Good job.

Reply:
Thank you for your review, i think you catched all the points i left behind ;)